### PR TITLE
Improve fees module

### PR DESCRIPTION
**Pull request by thiolliere**
related to https://github.com/paritytech/substrate/issues/1819

Note: for some reasons we can't add SimpleArithmetic bound to TransferAsset::Amount:
```diff
diff --git a/core/sr-primitives/src/traits.rs b/core/sr-primitives/src/traits.rs
index 5e12d571..ddb8a867 100644
--- a/core/sr-primitives/src/traits.rs
+++ b/core/sr-primitives/src/traits.rs
@@ -148,7 +148,7 @@ pub trait ChargeFee<AccountId>: ChargeBytesFee<AccountId> {
 /// Transfer fungible asset trait
 pub trait TransferAsset<AccountId> {
        /// The type of asset amount.
-       type Amount;
+       type Amount: SimpleArithmetic;
 
        /// Transfer asset from `from` account to `to` account with `amount` of asset.
        fn transfer(from: &AccountId, to: &AccountId, amount: Self::Amount) -> Result<(), &'static str>;
```
result in
```
error[E0275]: overflow evaluating the requirement `_: std::marker::Sized`
   --> core/sr-primitives/src/traits.rs:174:9
    |
174 | impl<T> TransferAsset<T> for () {
    |         ^^^^^^^^^^^^^^^^
    |
    = help: consider adding a `#![recursion_limit="128"]` attribute to your crate
    = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Permill::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Permill::_parity_codec::Compact<_>`
    = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Permill::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Permill::_parity_codec::Compact<_>`
    = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Permill::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Permill::_parity_codec::Compact<_>`
    = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Permill::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Permill::_parity_codec::Compact<()>`
    = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Permill::_parity_codec::HasCompact` for `()`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0275`.
error: Could not compile `sr-primitives`.

To learn more, run the command again with --verbose.
```

So and also I think it is better, I went for the second option:
I made a trait ArithmeticType for bound its associated type to some arithmetic stuff and else, and make balances::Module implementing it.
```rust
pub trait ArithmeticType {
	type Type: SimpleArithmetic + As<usize> + As<u64> + Codec + Copy + MaybeSerializeDebug + Default;
}
```

Also while doing so I run into an error of the bound of serde by `get_non_bound_serde_derive_types` basically types like AssetOf<T> bound T to serialize explicitly in our implementation and also `<<T as Trait>::TransferAsset as ArithmeticType::Type` doesn't works.
For this part I wonder why we try to be clever. Rust handle where clause very well already binding all types seems fine. AFAIK it is what we have done for parity-codec-derive also.

Last thing. I bound some arithmetic stuff to the associated type Balance of Currency in the past. It seems we should use this ArithmeticType instead so I replaced in modules.

### CODE REVIEW

**Comment by thiolliere** [comment]
for the sake of clarity, now modules like treasury bounds its associated type like 
`type Currency: ArithmeticType + Currency<Self::AccountId, Balance=BalanceOf<Self>>;
`

File: node/runtime/src/lib.rs [hunk 1]
```diff
@@ -60,8 +60,8 @@ pub const VERSION: RuntimeVersion = RuntimeVersion {
 	spec_name: create_runtime_str!("node"),
 	impl_name: create_runtime_str!("substrate-node"),
 	authoring_version: 10,
-	spec_version: 29,
```
Comment (gavofyork) [inline node/runtime/src/lib.rs]: no need to bump spec version - there are no logical changes.

**Comment by gavofyork** [comment]
Minor change but otherwise good to go.

**Comment by bkchr** [comment]
```
error[E0275]: overflow evaluating the requirement `_: std::marker::Sized` --> core/sr-primitives/src/traits.rs:174:9 | 174 | impl<T> TransferAsset<T> for () { | ^^^^^^^^^^^^^^^^ | = help: consider adding a `#![recursion_limit="128"]` attribute to your crate = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Permill::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Permill::_parity_codec::Compact<_>` = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Permill::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Permill::_parity_codec::Compact<_>` = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Permill::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Permill::_parity_codec::Compact<_>` = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Permill::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Permill::_parity_codec::Compact<()>` = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Permill::_parity_codec::HasCompact` for `()` error: aborting due to previous error For more information about this error, try `rustc --explain E0275`. error: Could not compile `sr-primitives`. To learn more, run the command again with --verbose.
```

Sounds like a bug in our parity codec derive logic. Would be nice if you could create some reproducible example and open an issue in parity-codec.

**Comment by thiolliere** [comment]
how did you get this error ?

I got a Sized issue when doing that. If I use Currency as the main trait bound and other I assign there inner type to the currency one then it fails to compiles. But if use ArithmeticType as the main trait bound and other I assigne to ArithmeticType::Type then it works.
a minimal not working example would be: (but I don't understand it at the moment but if you think this could be solved in parity-codec I can think deeper on it)
```rust
type Of<T> = <<T as F>::Type as Cur>::Type;
pub trait Arith {
	type Type: parity_codec::HasCompact;
}
pub trait Cur {
	type Type;
}
pub trait F {
	type Type: Cur + Arith<Type = Of<Self>>;
}
```
error is:
```stdout
    Checking srml-treasury v0.1.0 (/home/thiolliere/Developpement/substrate/srml/treasury)
error[E0275]: overflow evaluating the requirement `_: std::marker::Sized`
  --> srml/treasury/src/lib.rs:39:1
   |
39 | / pub trait F {
40 | |     type Type: Cur + Arith<Type = Of<Self>>;
41 | | }
   | |_^
   |
   = help: consider adding a `#![recursion_limit="128"]` attribute to your crate
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<<<Self as F>::Type as Cur>::Type>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::HasCompact` for `<<Self as F>::Type as Cur>::Type`
note: required by `Arith`
  --> srml/treasury/src/lib.rs:33:1
   |
33 | pub trait Arith {
   | ^^^^^^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0275`.
error: Could not compile `srml-treasury`.

To learn more, run the command again with --verbose.
```

**Comment by thiolliere** [comment]
```rust
pub trait F {
	type T: parity_codec::HasCompact;
}

impl F for () {
	type T = ();
}
```
```
   |
36 | impl F for () {
   |      ^
   |
   = help: consider adding a `#![recursion_limit="128"]` attribute to your crate
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`

   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<_>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::Decode` for `_IMPL_DECODE_FOR_Proposal::_parity_codec::Compact<()>`
   = note: required because of the requirements on the impl of `_IMPL_DECODE_FOR_Proposal::_parity_codec::HasCompact` for `()`
```

should we simply implement Compact<()> ?

**Comment by thiolliere** [comment]
OK so 
```diff
diff --git a/src/codec.rs b/src/codec.rs
index 730640d..a44d61a 100644
--- a/src/codec.rs
+++ b/src/codec.rs
@@ -213,7 +213,7 @@ macro_rules! impl_from_compact {
        }
 }
 
-impl_from_compact! { u8, u16, u32, u64, u128 }
+impl_from_compact! { (), u8, u16, u32, u64, u128 }
 
 /// Compact-encoded variant of &'a T. This is more space-efficient but less compute-efficient.
 #[derive(Eq, PartialEq, Clone, Copy)]
@@ -277,6 +277,18 @@ impl<T: 'static> HasCompact for T where
        type Type = Compact<T>;
 }
 
+impl<'a> Encode for CompactRef<'a, ()> {
+       fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R {
+               f(&[])
+       }
+}
+
+impl Decode for Compact<()> {
+       fn decode<I: Input>(_input: &mut I) -> Option<Self> {
+               Some(Compact(()))
+       }
+}
+
 // compact encoding:
 // 0b00 00 00 00 / 00 00 00 00 / 00 00 00 00 / 00 00 00 00
 //   xx xx xx 00                                                                                                                       (0 ... 2**6 - 1)                (u8)
```
solve this recursive error.

TransferAsset can't bound SimpleArithmetic then because () doesn't implements Mul Add and things but at least error is understandable. and it can bound HasCompact and it compiles.

Also if we do that we should also implement Encode for `Compact<()>`.

**Comment by thiolliere** [comment]
seems fine to me @bkchr

**Comment by bkchr** [comment]
Okay :) Thanks for looking into the error again!

**Pull request merged**